### ISSUE TITLE
Eml attachments in Outlook messages

### ISSUE THREAD

**Issue by dave0292** [jstedfast/MimeKit#677]
Hi Jeff, 

  I have been able to parse an Outlook message that it seems to be in MIME format (Outlook connects to my hMail server where I inspect the content of the emails). But when I add an eml attachment with more attachments inside, I can not access to them. The MimePart with the "Attachment.eml" filename is detected but is not parsed as a MessagePart. Im almost sure that im doing something wrong, but I do not understand why I can not extract the attachments from the eml if the Outlook message is correctly loaded with Mimekit. Im using MimeIterator to get all the parts and the ".eml" attachment is the last one (I should be able to access to documents inside of the ".eml").

  Moreover, when I send an Outlook message with nested attachments in ".msg" format I can not obtain documents inside the attachment from the third level, I just get the "Third level.msg" as a MimePart as in the previous case with the eml. 

Thanks, 

  David.

**Comment by jstedfast** [comment]
I would have to see the raw message to know why you are having issue.

But... here's the first question I'd be asking:

What is the `Content-Type` header for these eml attachments? Is it `message/rfc822`?

**Comment by dave0292** [comment]
If I send an Outlook message with the following structure, Outlook adds the content-type "{Content-Type: application/octet-stream; name="Second level Email.eml"}" to the second level attached email "Second level Email.eml", and I can not extract any files from there: 

Main message:
   - FirstLevel.docx.
   - Second level Email.eml.
Second level Email.eml:
   - Second level.pptx.
   - Third level Email.eml.
Third level  Email.eml:
   - ThirdLevel.txt.

But if I inspect the email file, I can see the second and third level attachments with the following format format:

------=_NextPart_000_0014_01D75223.047C31A0
Content-Type: application/octet-stream;
	name="Second level Email.eml"
Content-Transfer-Encoding: quoted-printable
Content-Disposition: attachment;
	filename="Second level Email.eml"

*******************************************************************
--------------3D93A9B83258CD706FD9BD49
Content-Type: message/rfc822;
 name=3D"Third level Email.eml"
Content-Transfer-Encoding: 7bit
Content-Disposition: attachment;
 filename=3D"Third level Email.eml"

Using Outlook and MimeIterator I'm getting only the first level attachment (FirstLevel.docx) and "Second level Email.eml" as MimePart (so I'm adding it wrongly to my list of attachments also). If I send this structure with ThunderBird everything works fine. 

I wrote this message yesterday, and today I can get all the attachments correctly with the same code (including attachments from another depth level). I don't see the reason or any different in the tests I made. But I have seen today that when I get all the attachments correctly, MessagePart's don't have a name as in the previous example. They just have the content type set as {Content-Type: message/rfc822} and they are null if they are parsed as MimeParts. 

I don't know if this gives you any hints but I'm pretty confused with this strange behavior. I'm trying with 2 different clients, one my host and a VM client with Microsoft 365 Outlook version. Could it be something related with the "Quoted-printable" Content-Transfer-Encoding and ".eml" attachments? Why I'm been able to load Outlook messages that are supposed to work with .msg format and some ".eml" attachments that are in Mime format (created with ThunderBird) are not correctly sent?

Kind regards, 

  David.

**Comment by jstedfast** [comment]
The problem is that something is setting the `Content-Type` to `application/octet-stream`. This is a generic "I don't know what the content is" Content-Type which means that MimeKit doesn't know that it's a message and so does not recursively try to parse it.

MimeKit will only recursively try to parse `message/rfc822` parts.

My guess is that Thunderbird is correctly setting the `Content-Type` to `message/rfc822`.

> Could it be something related with the "Quoted-printable" Content-Transfer-Encoding and ".eml" attachments?

It's also not allowed to encode a message/rfc822 with quoted-printable or base64 since it would make it impossible for the parser to parse the content. It would have to first decode the entire raw content and then parse but parsers are meant to parse w/o having to decode first.

**Comment by dave0292** [comment]
And is there a way to parse an attached email to MessagePart, or to try it at least when the filename extension is .msg or .eml? It can not be done simply doing "entity as MessagePart" as the result is null.
I have tried again with nested .msg files and the content-type of an attached email is the following:

{Content-Type: application/octet-stream; name="First level protected email.msg"}

I think that sometimes works correctly and uses the correct content type and others not.

**Comment by jstedfast** [comment]
What you can do is:

```csharp
if (mimePart.FileName != null && mimePart.FileName.EndsWith (".eml")) {
    using (var contentStream = mimePart.Content.Open ()) {
        var message = MimeMessage.Load (contentStream);
    }
}
```

**Comment by dave0292** [comment]
This code worked great, but I have found issues in two cases:

- When trying to access nested ".msg" attachments from third level. 
- Accessing ".msg" (with the previous code) attachments obtained from a Tnef part. The Content-Type is "{Content-Type: application/octet-stream; name="Level 3.msg"}" as in the previous case.

The error is the same in both cases, "Failed to parse message headers.". I think that the error is created in "StepHeaders" (line 861), but being a format error, I don't know if there is a way of fixing this.

Thanks in advance.

**Comment by jstedfast** [comment]
I would have to see the raw messages to figure out why there is a format error to know for sure.

If you can find an example that doesn't contain sensitive info (or if you find one where you can easily X-out the sensitive details w/o changing the overall MIME structure), you can either post it here in a comment or zip it up into a .zip file and send it to me and I can take a look (I need it in .zip format to make sure Outlook isn't munging it in any way when I save the attachment).

**Comment by dave0292** [comment]
Hi Jeff, I wrote you to "jestedfa@microsoft.com". Let me know if you do not receive the email. Thanks.

**Comment by jstedfast** [comment]
Yep, I just got your message (and just saw your message from Thursday as well). Sorry I didn't see your previous message until now - I was out of town last week.

**Comment by jstedfast** [comment]
Oddly, the message that you sent me loads fine?

**Comment by jstedfast** [comment]
Oh, I see... I thought the content you sent was the content of the level2 message, but it *contains* the .msg attachment.

**Comment by jstedfast** [comment]
Okay, so the issue is that the ".msg" attachment is not a MIME message. It's actually TNEF data.

So what you need to do is this:

```csharp
if (mimePart.FileName.EndsWith (".msg")) {
    using (var contentStream = new MemoryStream ()) {
        mimePart.Content.DecodeTo (contentStream);
        contentStream.Position = 0;

        var tnef = new TnefPart () {
            Content = new MimeContent (contentStream)
        };

        var message = tnef.ConvertToMessage ();
    }
} else if (mimePart.FileName.EndsWith (".eml")) {
    using (var contentStream = mimePart.Content.Open ()) {
        var message = MimeMessage.Load (contentStream);
    }
}
```

**Comment by dave0292** [comment]
Hi Jeff, the code works fine when loading the message, but I'm not able to extract the attachments from this TnefPart: 
``` 
if (part.FileName != null &&
  (part.FileName.EndsWith(".eml") || part.FileName.EndsWith(".msg")))
  {
      MimeMessage nestedMessage = null;
      
      if (part.FileName.EndsWith(".msg"))
      {
            using (var contentStream = new MemoryStream())
            {
                    part.Content.DecodeTo(contentStream);
                    contentStream.Position = 0;
                
                    var tnef = new TnefPart()
                    {
                        Content = new MimeContent(contentStream)
                    };
                
                    IEnumerable<MimeEntity> tnefAttachments = tnef.ExtractAttachments();
                    foreach (MimeEntity tnefMimeEntity in tnefAttachments.OfType<MimePart>())
                    {
                        
                    }
          
                  nestedMessage = tnef.ConvertToMessage();
            }
      }
      else if (part.FileName.EndsWith(".eml"))
      {
              using (var contentStream = part.Content.Open())
              {
                  nestedMessage = MimeMessage.Load(contentStream);
              }
      }
      
      //make sure to reset the stream
      //contentStream.Position = 0;
      
      //MimeMessage nestedMessage = MimeMessage.Load(contentStream);
      var iterator = new MimeIterator(nestedMessage);
      
      while (iterator.MoveNext())
      {
      var parentMimePart = iterator.Parent as Multipart;
      var currentMimePart = iterator.Current as MimeEntity;
      int nestedDepth = iterator.Depth;
      
      if (parentMimePart != null && currentMimePart != null && currentMimePart.IsAttachment)
                ExtractAttachmentsFromMimeEntity(currentMimePart, parentMimePart, nestedDepth + depth + 1, attachmentsFromMessage);
  }
```

I have used the TnefPart ExtractAttachments method and the iterator after converting it to a MimeMessage, but both of them give the same result: they return just one part that is not an attachment. 

Best regards, 

  David.

**Comment by jstedfast** [comment]
Okay, odds are that the TNEF parser will need changes.

**Comment by dave0292** [comment]
Hi Jeff, 

I have seen that the reason of  not being able to load the ".msg" and extract the attachments from the inside is that I was adding the ".msg" file directly to the "attachment box" (that is created when one or more attachments are added) instead of using the "drag&drop" to add it to the body of the email. Adding the message using the insert button would also add the nested attachment. 

When the ".msg" message is added with the correct format Outlook adds a label under the file with the value "Outlook item". If you drag a message to the place holder where other attachments are added the format is not correctly added and you can see the ".msg" extension (while adding it with the button or the the body of the message hides the extension as it is recognized). 

I will try to extract attachments from those raw files, but I'm not sure if it will be possible. With your previous advices I was able to load the nested message without errors but it does not contain attachments and the MimeIterator obtained from that nested message only sees one part. 

Thanks for your help.

Best regards, 

David.

**Comment by jstedfast** [comment]
This may be related to issue #685

**Comment by jstedfast** [comment]
@dave0292 I dug into this some more and the "Second level message .msg" attachment isn't valid TNEF - it's something else.

**Comment by dave0292** [comment]
Hi Jeff, 

Thanks for the info. I have not been to figure out how to do this. I have also seen that this format error also appears when using Outlook web and adding a nested ".msg". There is not way to add the nested attachment with the correct format in the web client, always adds it as a raw file.

Best regards, 

David.

**Comment by jstedfast** [comment]
Okay, so that .msg attachment is probably the Outlook MSG format which this library can likely parse: https://github.com/Sicos1977/MsgKit

**Comment by dave0292** [comment]
I tried to load a ".msg" message sent from Outlook web with the content-type set to "application/octet-stream" and Aspose.Email is not able to load it as a "MapiMessage", so I think its not the Outlook format. Unfortunately, I checked the code of "MsgKit" and I think it's not able to read an incoming message (as far as I knkow), its just possible to create new messages and appointments from 0. 
I think I will need to use something like Aspose.Mail. Thank you very much for your help Jeff.

**Comment by jstedfast** [comment]
@dave0292 ah, I guess I just assumed MsgKit could parse them as well. Sorry about that.

Good luck with whatever route you go.

**Comment by dave0292** [comment]
I have tried to parse this type of message (which can be easily reproduced adding an Outlook .msg message with an attachment and sending it with Outlook Web client) as a MapiMessage (Outlook message in Aspose Email) and it seems that is not correct format either. 

The attachment headers start like this: 

--_004_DB8PR10MB3018FFBA5889FB9C9F287BC6F8199DB8PR10MB3018EURP_
Content-Type: application/octet-stream; name="Attachment inside.msg"
Content-Description: Attachment inside.msg
Content-Disposition: attachment; filename="Attachment inside.msg";
	size=105472; creation-date="Thu, 08 Jul 2021 08:10:24 GMT";
	modification-date="Thu, 08 Jul 2021 08:10:38 GMT"
Content-Transfer-Encoding: base64

**0M8R4KGxGuEAAAAAAAAAAAAAAAAAAAAAPgADAP7/CQAGAAAAAAAAAAAAAAACAAAAAgAAAAAAAAAA
EAAAKQAAAAcAAAD+////AAAAAAMAAAB6AAAA////////////////////////////////////////
////////////////////////////////////////////////////////////////////////////...**

****************************
I know it has not the correct Mime headers, it should be marked as "rfc822", but for some reason Outlook is adding ".msg" attachments without the correct format, but there must be a way to extract attachments from there.

I have seen that if you decode from Base64 there are some MimeHeaders in the middle of the decoded content, but there are also some incorrect characters that could make the parser to fail when loading the message. Maybe there are two nested encodings?

Here you have the complete email extracted with EmailRelay: 
[EmailWithNestedMsgAndAttachment.txt](https://github.com/jstedfast/MimeKit/files/6783854/EmailWithNestedMsgAndAttachment.txt)

Here is also decoded (from Base64 with UTF8) the content of the ".msg" attachment: 
[DecodedFromBase64NestedAttachment.txt](https://github.com/jstedfast/MimeKit/files/6783891/DecodedFromBase64NestedAttachment.txt)


Best regards, 

David.

**Comment by dave0292** [comment]
Sorry Jeff, I think I found a solution. It seems that the content of the ".msg" attachment is encoded with Base64, so I decoded it and loaded as an Msg with MapiMessage class from Aspose. I need to modify the content of the attachments, that's why I could not use MsgReader (I think MsgKit does not allow either to read an already created Outlook message, just create it from 0).
Because of the encoding I was not being able to load the message as an Outlook message and I did not know if the problem was in my "Mime" code or the Outlook implementation. 
I hope this thread helps other people with this issue. 
Thank you very much :)
David.
